3_harmonize/src/clean_wqp_data.R

```diff
@@ -3,29 +3,103 @@
+                                                    'MonitoringLocationIdentifier',
+                                                    'ActivityStartDate', 
+                                                    'ActivityStartTime.Time',
+                                                    'CharacteristicName', 
+                                                    'ResultSampleFractionText')){
 
   # Clean data and assign flags if applicable
   wqp_data_clean <- wqp_data %>%
     # harmonize characteristic names by assigning a common parameter name
     # to the groups of characteristics supplied in `char_names_crosswalk`.
-    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name"))
+    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name")) %>%
+    # flag true missing results, i.e. when result measure value and detection
+    # limit value are both NA, when "not reported" is found in the column
+    # "ResultDetectionConditionText", or when any of the strings from
+    # `commenttext_missing` are found in the column "ResultCommentText".
+    mutate(flag_missing_result = case_when(
+      is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ~ TRUE,
+      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) ~ TRUE,
+      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE) ~ TRUE
+    )) %>%
+    # Flag duplicate records
+    flag_duplicates(., duplicate_definition)
   
   return(wqp_data_clean)
   
   
 }
 
+
+#' @title Flag duplicated records
+#' 
+#' @description 
+#' Function to flag duplicated rows based on a user-supplied definition
+#' of a duplicate record. Additional flags can help reconcile duplicate sets.
+#' For example, a flag "flag_duplicate_drop_random" is assigned based on 
+#' randomly selecting the first record from a duplicated set, making it 
+#' easier to drop all others from the dataset. 
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a unique data record.
+#' @param duplicate_definition character string(s) indicating which columns are
+#' used to identify a duplicate record. Duplicate records are defined as those 
+#' that share the same value for each column within `duplicate_definition`.
+#'
+#' @returns 
+#' Returns a data frame containing data downloaded from the Water Quality Portal,
+#' where each row represents a unique data record. New columns appended to the 
+#' original data frame include flags for duplicated records. 
+#' 
+flag_duplicates <- function(wqp_data, duplicate_definition){
+  
+  wqp_data_out <- wqp_data %>%
+    # Step 1: Flag duplicate records using the `duplicate_definition`
+    group_by(across(all_of(duplicate_definition))) %>% 
+    mutate(n_duplicated = n(),
+           flag_duplicated_row = if_else(n_duplicated > 1, TRUE, NA)) %>% 
```

Comment:
I am actually thinking that we could break this into two functions: `flag_duplicates()` which does your `Step 1` here and `remove_duplicates()` which does your `Step 2` here but actually adds a `filter()`. Then, `clean_wqp_data()` can call them both and people could easily turn off the `remove_duplicates()` step or fairly easily modify its logic.

Reply:
OK, what do you think of this? I've moved the code in `step 2` above to its own function called `remove_duplicates()` in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/74/commits/16a02f96399596369cd685460746868b4a7cf178. I also added an argument to `clean_wqp_data()` to make it easier for someone to either retain or drop the extraneous duplicates in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/74/commits/0ad3f196de9ed953f11a86c51d105ba348aa9645. I think that my decisions are swayed by example codes I've seen that rely heavily on flagging (but retaining) everything rather than just omitting rows. So I could be convinced that this extra option is unnecessary.

Reply:
I like the flexibility of this additional argument. I think some people are uncomfortable with removing them (I never am ...) so it is a nice way to handle it.

Reply:
I also prefer to remove them. I do like that the added message highlights how many duplicated records there can be in a WQP dataset! Sometimes I wonder where they all come from, i.e. when STORET migrated to WQX did some records get duplicated?